### PR TITLE
Feat/assertion 2

### PR DESCRIPTION
**Pull request by kenanfarukcakir**

### CODE REVIEW

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 1.38 / 15
Max. Delta Alloc op: 1.64 / 15
Max. Delta Allocs op: 2.96 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 3.57 / 15
Max. Delta Alloc op: 2.38 / 15
Max. Delta Allocs op: 4.08 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 5.99 / 15
Max. Delta Alloc op: 3.46 / 15
Max. Delta Allocs op: 5.09 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 0.47 / 15
Max. Delta Alloc op: 0.44 / 15
Max. Delta Allocs op: 0.10 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 0.85 / 15
Max. Delta Alloc op: 0.26 / 15
Max. Delta Allocs op: 0.67 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: -0.00 / 15
Max. Delta Alloc op: 0.17 / 15
Max. Delta Allocs op: 0.12 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 2.31 / 15
Max. Delta Alloc op: 0.16 / 15
Max. Delta Allocs op: 0.72 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 1.50 / 15
Max. Delta Alloc op: 0.61 / 15
Max. Delta Allocs op: 2.02 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 34.55 / 15
Max. Delta Alloc op: 4.36 / 15
Max. Delta Allocs op: 3.95 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 1.27 / 15
Max. Delta Alloc op: 6.35 / 15
Max. Delta Allocs op: 8.28 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: -0.00 / 15
Max. Delta Alloc op: 1.05 / 15
Max. Delta Allocs op: 0.52 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 7.43 / 15
Max. Delta Alloc op: 5.26 / 15
Max. Delta Allocs op: 8.24 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 0.00 / 15
Max. Delta Alloc op: 3.12 / 15
Max. Delta Allocs op: 0.33 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 1.18 / 15
Max. Delta Alloc op: 1.80 / 15
Max. Delta Allocs op: 1.73 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 0.01 / 15
Max. Delta Alloc op: 2.29 / 15
Max. Delta Allocs op: 3.19 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: -0.00 / 15
Max. Delta Alloc op: 1.62 / 15
Max. Delta Allocs op: 1.96 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 1.76 / 15
Max. Delta Alloc op: 1.76 / 15
Max. Delta Allocs op: 2.43 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 2.77 / 15
Max. Delta Alloc op: 0.82 / 15
Max. Delta Allocs op: 1.84 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 5.48 / 15
Max. Delta Alloc op: 0.86 / 15
Max. Delta Allocs op: 1.17 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 0.88 / 15
Max. Delta Alloc op: 0.55 / 15
Max. Delta Allocs op: 0.66 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 13.17 / 15
Max. Delta Alloc op: 11.95 / 15
Max. Delta Allocs op: 17.07 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 2.28 / 15
Max. Delta Alloc op: 0.31 / 15
Max. Delta Allocs op: 0.66 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 2.26 / 15
Max. Delta Alloc op: 1.47 / 15
Max. Delta Allocs op: 2.15 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 11.28 / 15
Max. Delta Alloc op: 7.55 / 15
Max. Delta Allocs op: 11.33 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 4.49 / 15
Max. Delta Alloc op: 4.69 / 15
Max. Delta Allocs op: 6.81 / 15

File: README.md [hunk 1]
```diff
@@ -320,7 +320,29 @@ There is an example config file at [config_examples/config.json](/config_example
     | `order`   | Order of reading records from CSV. Can be `random` or `sequential`                                | `string`    | `random`    | No         |
     | `skip_first_line`   | Skips first line while reading records from CSV.                                | `bool`    | `false`    | No         |
     | `skip_empty_line`   | Skips empty lines while reading records from CSV.                                | `bool`    | `true`    | No         |
-   
+
+- `assertions` *optional*
+    Config for test-wide assertion. *abort* and *delay* fields can be used to adjust the abort behaviour in case assertion failure. If abort is true for a rule and rules fails at certain point, engine will decide to abort test immediately if delay is 0 or not given. If delay is given, it will wait for delay seconds and reassert the rule.
+    **Example:** Check *90th percentile* and *fail_count*;
+    ```json
+     "assertions": [
```
Comment (kursataktas) [inline README.md:328]: let's show the user where this config blog is located.

```json
{
  "duration": 10,
   <other_global_configurations>,
  "assertions": [
      {
        "rule" : "p90(iteration_time) < 220", 
        "abort" : false
      },
      {
        "rule" : "fail_count_perc < 0.1", 
        "abort" : true,
        "delay" : 1
      },
      {
        "rule" : "fail_count < 100", 
        "abort" : true,
        "delay" : 0
      }
   ],
   steps: [
        ....
   ],
}
```

File: README.md [hunk 2]
```diff
@@ -671,7 +693,41 @@ If Ddosify can't receive the response for a request, that step is marked as Fail
 | `(status_code == 200) \|\| (status_code == 201)`   | same as preceding one |
 | `regexp(body,\"[a-z]+_[0-9]+\",0) == \"messi_10\"`   | checks if matched result from regex is equal to "messi_10" |
 
+## Test-Wide Assertion
+
+Ddosify now supports test-wide assertions, allowing users to verify the success of their load tests based on response times and failure counts of iterations. With this feature, users can assert the percentile of response times and the failure counts of all the iterations in a test.
```
Comment (kursataktas) [inline README.md:698]: - Ddosify now supports -> Let's remove _**now**_

- Let's put the below paragraph at the end of this section.

**Difference Between Test-Wide Assertions and Step Assertions**
Unlike assertions focused on individual steps, which determine the success or failure of a step according to its response, Test-Wide assertions create an abort/continue logic for the entire test, which is based on the accumulated data from all iterations.

File: README.md [hunk 3]
```diff
@@ -671,7 +693,41 @@ If Ddosify can't receive the response for a request, that step is marked as Fail
...
+
+Users can specify the required percentile of response times and failure counts in the configuration file, and the engine will compare the actual response times and failure counts to these values throughout the test continuously. According to users configuration test can be aborted or continue running until end. Check the [example config](https://github.com/ddosify/ddosify/blob/master/config_examples/config.json) to see how it looks.
+
+Note that functions and operators mentioned in Assertion section can be used test-wide assertions also.
```
Comment (kursataktas) [inline README.md:702]: Note that functions and operators mentioned in Assertion section can be used test-wide assertions also.

⬇️ 

Note that the functions and operators mentioned in the [Step Assertion](https://github.com/ddosify/ddosify#assertion) section can also be utilized for the Test-Wide Assertion keywords listed below.

**Review by kursataktas — COMMENTED** [review]
🚀

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 0.54 / 15
Max. Delta Alloc op: 0.41 / 15
Max. Delta Allocs op: 0.59 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 0.01 / 15
Max. Delta Alloc op: 0.51 / 15
Max. Delta Allocs op: 0.97 / 15

**Comment by ddosifyadmin** [comment]
Max. Delta Time op: 3.52 / 15
Max. Delta Alloc op: 0.78 / 15
Max. Delta Allocs op: 0.84 / 15

**Pull request merged**